File: mahjong.py
from copy import deepcopy
import pickle
from pathlib import Path
import re


SUITS = "条饼万东西南北中发白"
NUMS = "一二三四五六七八九"
TILES_UNICODE = [
    ["🀐", "🀑", "🀒", "🀓", "🀔", "🀕", "🀖", "🀗", "🀘"],
    ["🀙", "🀚", "🀛", "🀜", "🀝", "🀞", "🀟", "🀠", "🀡"],
    ["🀇", "🀈", "🀉", "🀊", "🀋", "🀌", "🀍", "🀎", "🀏"],
    ["🀀"],
    ["🀂"],
    ["🀁"],
    ["🀃"],
    ["🀄"],
    ["🀅"],
    ["🀆"],
]
TILES_TEXT = [
    ["一条", "二条", "三条", "四条", "五条", "六条", "七条", "八条", "九条"],
    ["一饼", "二饼", "三饼", "四饼", "五饼", "六饼", "七饼", "八饼", "九饼"],
    ["一万", "二万", "三万", "四万", "五万", "六万", "七万", "八万", "九万"],
    ["东"],
    ["西"],
    ["南"],
    ["北"],
    ["中"],
    ["发"],
    ["白"],
]


def hand_count(hand):
    return sum(sum(s) for s in hand)


def print_hand(hand, unicode=False, end="\n"):
    for s in range(10):
        for rank, count in enumerate(hand[s]):
            for _ in range(count):
                if unicode:
                    print(TILES_UNICODE[s][rank], end=" ")
                else:
                    print(TILES_TEXT[s][rank], end=" ")
    print()


def parse_hand(str):
    tiles = re.split(r"\s+", str.strip())
    hand = [[0] * 9 for _ in range(10)]
    for tile in tiles:
        if len(tile) == 1:
            s = SUITS.index(tile)
            hand[s][0] += 1
        else:
            s = SUITS.index(tile[1])
            n = NUMS.index(tile[0])
            hand[s][n] += 1
    return hand


def parse_tile(str):
    if len(str) == 1:
        return SUITS.index(str), 0
    else:
        return SUITS.index(str[1]), NUMS.index(str[0])


def init_tables():
    def gen3n(a, level, table):
        if level > 4:
            return
        # add a triplet
        for i in range(9):
            if a[i] + 3 > 4:
                continue
            b = a.copy()
            b[i] += 3
            table.add(tuple(b))
            gen3n(b, level + 1, table)
        # add a sequence
        for i in range(7):
            if (a[i] + 1 > 4) or (a[i + 1] + 1 > 4) or (a[i + 2] + 1 > 4):
                continue
            b = a.copy()
            b[i] += 1
            b[i + 1] += 1
            b[i + 2] += 1
            table.add(tuple(b))
            gen3n(b, level + 1, table)

    def gen3n2(table):
        for i in range(9):
            a = [0] * 9
            a[i] += 2
            table.add(tuple(a))
            gen3n(a, 1, table)

    global TAB3N, TAB3N2
    path = Path(__file__).parent / "win.pickle"
    if path.exists():
        with open(path, "rb") as f:
            TAB3N, TAB3N2 = pickle.load(f)
    else:
        TAB3N = set()
        TAB3N2 = set()
        TAB3N.add((0,) * 9)
        gen3n([0] * 9, 1, TAB3N)
        gen3n2(TAB3N2)
        with open(path, "wb") as f:
            pickle.dump((TAB3N, TAB3N2), f)


def checkwin(hand):
    if hand_count(hand) != 14:
        return False
    have3n2 = False
    for s in range(10):
        a = tuple(hand[s])
        if a in TAB3N2:
            if have3n2:
                return False
            else:
                have3n2 = True
        elif a not in TAB3N:
            return False
    return True


def wintile(hand):
    if hand_count(hand) != 13:
        return []
    tiles = []
    # 条 饼 万
    for s in range(3):
        for rank in range(9):
            if hand[s][rank] >= 4:
                continue
            hand_copy = deepcopy(hand)
            hand_copy[s][rank] += 1
            if checkwin(hand_copy):
                tiles.append(TILES_TEXT[s][rank])
    # 东 西 南 北 中 发 白
    for s in range(3, 10):
        if hand[s][rank] >= 4:
            continue
        hand_copy = deepcopy(hand)
        hand_copy[s][0] += 1
        if checkwin(hand_copy):
            tiles.append(TILES_TEXT[s][0])
    return tiles


def suggest(hand):
    if hand_count(hand) != 14:
        return []
    tiles = {}
    for s in range(10):
        for rank in range(9):
            if hand[s][rank] < 1:
                continue
            hand_copy = deepcopy(hand)
            hand_copy[s][rank] -= 1
            win_tiles = wintile(hand_copy)
            if len(win_tiles) > 0:
                info = []
                for tile in win_tiles:
                    win_tile_suit, win_tile_rank = parse_tile(tile)
                    info.append((tile, 4 - hand_copy[win_tile_suit][win_tile_rank]))
                tiles[TILES_TEXT[s][rank]] = info
    return tiles


init_tables()

File: docs/mahjong.js
const SUITS = '条饼万东西南北中发白';
const NUMS = '一二三四五六七八九';

const TILES_TEXT = [
  ['一条', '二条', '三条', '四条', '五条', '六条', '七条', '八条', '九条'],
  ['一饼', '二饼', '三饼', '四饼', '五饼', '六饼', '七饼', '八饼', '九饼'],
  ['一万', '二万', '三万', '四万', '五万', '六万', '七万', '八万', '九万'],
  ['东'],
  ['西'],
  ['南'],
  ['北'],
  ['中'],
  ['发'],
  ['白'],
];

function handCount(hand) {
  return hand.reduce((sum, suit) => sum + suit.reduce((suitSum, count) => suitSum + count, 0), 0);
}

function parseTile(str) {
  if (str.length === 1) {
    return [SUITS.indexOf(str), 0];
  } else {
    return [SUITS.indexOf(str[1]), NUMS.indexOf(str[0])];
  }
}

let TAB3N, TAB3N2;

function gen3n(a, level, table) {
  if (level > 4) {
    return;
  }
  // add a triplet
  for (let i = 0; i < 9; i++) {
    if (a[i] + 3 > 4) {
      continue;
    }
    const b = a.slice();
    b[i] += 3;
    table.add(b.join(','));
    gen3n(b, level + 1, table);
  }
  // add a sequence
  for (let i = 0; i < 7; i++) {
    if (a[i] + 1 > 4 || a[i + 1] + 1 > 4 || a[i + 2] + 1 > 4) {
      continue;
    }
    const b = a.slice();
    b[i] += 1;
    b[i + 1] += 1;
    b[i + 2] += 1;
    table.add(b.join(','));
    gen3n(b, level + 1, table);
  }
}

function gen3n2(table) {
  for (let i = 0; i < 9; i++) {
    const a = Array(9).fill(0);
    a[i] += 2;
    table.add(a.join(','));
    gen3n(a, 1, table);
  }
}

function initTables() {
  const cachedTab3N = localStorage.getItem('TAB3N');
  const cachedTab3N2 = localStorage.getItem('TAB3N2');

  if (cachedTab3N && cachedTab3N2) {
    TAB3N = new Set(JSON.parse(cachedTab3N));
    TAB3N2 = new Set(JSON.parse(cachedTab3N2));
  } else {
    TAB3N = new Set();
    TAB3N2 = new Set();
    TAB3N.add(Array(9).fill(0).join(','));
    gen3n(Array(9).fill(0), 1, TAB3N);
    gen3n2(TAB3N2);

    localStorage.setItem('TAB3N', JSON.stringify(Array.from(TAB3N)));
    localStorage.setItem('TAB3N2', JSON.stringify(Array.from(TAB3N2)));
  }
}

function checkWin(hand) {
  let have3n2 = false;
  for (let s = 0; s < 10; s++) {
    const a = hand[s].join(',');
    if (TAB3N2.has(a)) {
      if (have3n2) {
        return false;
      } else {
        have3n2 = true;
      }
    } else if (!TAB3N.has(a)) {
      return false;
    }
  }
  return true;
}

initTables();

function winTile(hand) {
  let tiles = [];
  // 条 饼 万
  for (let s = 0; s < 3; s++) {
    for (let rank = 0; rank < 9; rank++) {
      if (hand[s][rank] >= 4) {
        continue;
      }
      let handCopy = hand.map((suit) => suit.slice());
      handCopy[s][rank] += 1;
      if (checkWin(handCopy)) {
        tiles.push([TILES_TEXT[s][rank], 4 - hand[s][rank]]);
      }
    }
  }
  // 东 西 南 北 中 发 白
  for (let s = 3; s < 10; s++) {
    if (hand[s][0] >= 4) {
      continue;
    }
    let handCopy = hand.map((suit) => suit.slice());
    handCopy[s][0] += 1;
    if (checkWin(handCopy)) {
      tiles.push([TILES_TEXT[s][0], 4 - hand[s][0]]);
    }
  }
  return tiles;
}

function suggest(hand) {
  let tiles = [];
  for (let s = 0; s < 10; s++) {
    for (let rank = 0; rank < 9; rank++) {
      if (hand[s][rank] < 1) {
        continue;
      }
      let handCopy = hand.map((suit) => suit.slice());
      handCopy[s][rank] -= 1;
      let winTiles = winTile(handCopy);
      if (winTiles.length > 0) {
        let count = 0;
        for (let i = 0; i < winTiles.length; i++) {
          count += winTiles[i][1];
        }
        tiles.push([TILES_TEXT[s][rank], count]);
      }
    }
  }
  tiles.sort((a, b) => b[1] - a[1]);
  return tiles;
}
